File: solutions/python/change/1/change.py
from collections import deque


def find_fewest_coins(coins: list[int], target: int) -> list[int]:
    if target < 0:
        raise ValueError("target can't be negative")
    if target == 0:
        return []
    coins_biggest_to_smallest = sorted(coins, reverse=True)
    coinsets: deque[list[int]] = deque([[]])
    visited_values = set()
    while coinsets:
        coinset_prev = coinsets.popleft()
        for coin in coins_biggest_to_smallest:
            coinset = coinset_prev + [coin]
            value = sum(coinset)
            if value == target:
                return sorted(coinset)
            if value < target and value not in visited_values:
                visited_values.add(value)
                coinsets.append(coinset)
    raise ValueError("can't make target with given coins")

File: solutions/haskell/change/1/src/Change.hs
{-# LANGUAGE LambdaCase #-}

module Change (findFewestCoins) where

import Control.Applicative ((<|>))
import Data.Function (fix)
import Data.Set qualified as Set

findFewestCoins :: Integer -> [Integer] -> Maybe [Integer]
findFewestCoins 0 _ = Just []
findFewestCoins target coins =
    fix
        ( \continue visited_values -> \case
            [] -> Nothing
            coinset : coinsets_to_check ->
                lookup target new_values_coinsets
                    <|> continue
                        (visited_values <> Set.fromList new_values)
                        (coinsets_to_check <> new_coinsets)
              where
                value = sum coinset
                (new_values, new_coinsets) = unzip new_values_coinsets
                new_values_coinsets =
                    [ (new_value, coin : coinset)
                    | coin <- coins
                    , let new_value = coin + value
                    , new_value <= target
                    , new_value `notElem` visited_values
                    ]
        )
        Set.empty
        [[]]
